sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2457,164 +2450,127 @@ class Analyzer(override val catalogManager: CatalogManager)
-                    needsPushDown += aliasOrder
-                    order.copy(child = aliasOrder)
-                  } else {
-                    needsPushDown += evaluated
-                    order.copy(child = evaluated.toAttribute)
-                  }
-                } else {
-                  order.copy(child = originalAggExprs(index).toAttribute)
-                }
+          Sort(newSortOrder, global, newChild)
+        })
+    }
+
+    def resolveExprsWithAggregate(
+        exprs: Seq[Expression],
+        agg: Aggregate): (Seq[NamedExpression], Seq[Expression]) = {
+      val aggregateExpressions = ArrayBuffer.empty[NamedExpression]
+      val transformed = exprs.map { e =>
+        // Try resolving the expression as though it is in the aggregate clause.
+        def resolveCol(input: Expression): Expression = {
+          resolveExpressionByPlanOutput(input, agg.child)
+        }
+        def resolveSubQuery(input: Expression): Expression = {
+          if (SubqueryExpression.hasSubquery(input)) {
+            val fake = Project(Alias(input, "fake")() :: Nil, agg.child)
+            ResolveSubquery(fake).asInstanceOf[Project].projectList.head.asInstanceOf[Alias].child
+          } else {
+            input
           }
-
-          val sortOrdersMap = unresolvedSortOrders
-            .map(new TreeNodeRef(_))
-            .zip(evaluatedOrderings)
-            .toMap
-          val finalSortOrders = sortOrder.map(s => sortOrdersMap.getOrElse(new TreeNodeRef(s), s))
-
-          // Since we don't rely on sort.resolved as the stop condition for this rule,
-          // we need to check this and prevent applying this rule multiple times
-          if (sortOrder == finalSortOrders) {
-            sort
+        }
+        val maybeResolved = resolveSubQuery(resolveCol(e))
+        if (maybeResolved.resolved && maybeResolved.references.subsetOf(agg.outputSet) &&
+          !containsAggregate(maybeResolved)) {
+          // The given expression is valid and doesn't need extra resolution.
+          maybeResolved
+        } else if (containsUnresolvedFunc(maybeResolved)) {
+          // The given expression has unresolved functions which may be aggregate functions and we
+          // need to wait for other rules to resolve the functions first.
+          maybeResolved
+        } else {
+          // Avoid adding an extra aggregate expression if it's already present in
+          // `agg.aggregateExpressions`.
+          val index = if (maybeResolved.resolved) {
+            agg.aggregateExpressions.indexWhere {
+              case Alias(child, _) => child semanticEquals maybeResolved
+              case other => other semanticEquals maybeResolved
+            }
           } else {
-            Project(aggregate.output,
-              Sort(finalSortOrders, global,
-                aggregate.copy(aggregateExpressions = originalAggExprs ++ needsPushDown)))
+            -1
+          }
+          if (index >= 0) {
```

Comment:
It looks we don't need this if condition?
```
          if (maybeResolved.resolved) {
            val index = agg.aggregateExpressions.indexWhere {
              case Alias(child, _) => child semanticEquals maybeResolved
              case other => other semanticEquals maybeResolved
            }
            agg.aggregateExpressions(index).toAttribute
          } else {
            buildAggExprList(maybeResolved, agg, aggregateExpressions)
          }
```
?

Reply:
`val index = ...` this may return -1 if no match is found.

Reply:
Ah, I see.